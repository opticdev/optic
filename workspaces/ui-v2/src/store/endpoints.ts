import {
  createSlice,
  createAsyncThunk,
  SerializedError,
} from '@reduxjs/toolkit';
import { IForkableSpectacle } from '@useoptic/spectacle';
import { AsyncStatus, IEndpoint } from '<src>/types';
import { findLongestStartingSubstring } from '<src>/utils';

export const AllEndpointsQuery = `{
  requests {
    id
    pathId
    absolutePathPattern
    absolutePathPatternWithParameterNames
    pathComponents {
      id
      name
      isParameterized
      contributions
    }
    method
    pathContributions
  }
}`;

export type EndpointQueryResults = {
  requests: {
    id: string;
    pathId: string;
    absolutePathPattern: string;
    absolutePathPatternWithParameterNames: string;
    pathComponents: {
      id: string;
      name: string;
      isParameterized: boolean;
      contributions: Record<string, string>;
    }[];
    method: string;
    pathContributions: Record<string, string>;
  }[];
};

export const endpointQueryResultsToJson = ({
  requests,
}: EndpointQueryResults) => {
  const commonStart =
    requests.length > 0
      ? findLongestStartingSubstring(
          requests.map((req) => req.absolutePathPatternWithParameterNames)
        )
      : '/';

  return requests.map((request) => ({
    pathId: request.pathId,
    method: request.method,
    fullPath: request.absolutePathPatternWithParameterNames,
    group: request.absolutePathPatternWithParameterNames
      .substring(commonStart.length)
      .split('/')[0],
    pathParameters: request.pathComponents.map((path) => ({
      id: path.id,
      name: path.name,
      isParameterized: path.isParameterized,
      description: path.contributions.description || '',
      endpointId: `${request.pathId}.${request.method}`,
    })),
    description: request.pathContributions.description || '',
    purpose: request.pathContributions.purpose || '',
  }));
};

const fetchEndpoints = createAsyncThunk<
  EndpointQueryResults,
  { spectacle: IForkableSpectacle }
>('FETCH_ENDPOINTS', async ({ spectacle }) => {
  const results = await spectacle.query<EndpointQueryResults>({
    query: AllEndpointsQuery,
    variables: {},
  });
  if (results.errors) {
    throw new Error();
  }
  return results.data!;
});

const initialState: {
  results: AsyncStatus<IEndpoint[], SerializedError>;
} = {
  results: {
    loading: true,
  },
};

const endpointsSlice = createSlice({
  name: 'endpoints',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(fetchEndpoints.fulfilled, (state, action) => {
      const results = action.payload;
      console.log(results);
      console.log(endpointQueryResultsToJson(results));

      state.results = {
        loading: false,
        data: endpointQueryResultsToJson(results),
      };
    });
    builder.addCase(fetchEndpoints.rejected, (state, action) => {
      state.results = {
        loading: false,
        error: action.error,
      };
    });
  },
});

export const actions = {
  fetchEndpoints,
};
export const reducer = endpointsSlice.reducer;
