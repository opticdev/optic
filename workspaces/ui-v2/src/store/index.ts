export * from './root';
export { actions as endpointActions } from './endpoints';
export { actions as documentationEditActions } from './documentationEdit';
export { actions as metadataActions } from './metadata';
export * as selectors from './selectors';
