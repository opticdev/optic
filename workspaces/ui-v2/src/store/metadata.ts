import {
  createAsyncThunk,
  createSlice,
  SerializedError,
} from '@reduxjs/toolkit';
import { v4 as uuidv4 } from 'uuid';

import { Client } from '@useoptic/cli-client';
import {
  IOpticConfigRepository,
  IForkableSpectacle,
} from '@useoptic/spectacle';
import { AsyncStatus } from '<src>/types';

const fetchMetadata = createAsyncThunk(
  'FETCH_METADATA',
  async ({
    configRepository,
    spectacle,
  }: {
    configRepository: IOpticConfigRepository;
    spectacle: IForkableSpectacle;
  }) => {
    const client = new Client('/api');

    const apiNamePromise = configRepository.getApiName().catch(() => '');
    const clientAgentPromise = client
      .getIdentity()
      .then((response) => response.json())
      .then(({ anonymousId }) => anonymousId)
      .catch(() => 'anon_id');
    const specMetadataPromise = spectacle
      .query<{
        metadata: {
          id: string;
        };
      }>({
        query: `{
        metadata {
          id
        }
      }`,
        variables: {},
      })
      .then((result) => {
        if (!result.data) {
          throw new Error('Could not fetch spec metadata');
        }
        return result.data.metadata.id;
      });

    const [apiName, clientAgent, specificationId] = await Promise.all([
      apiNamePromise,
      clientAgentPromise,
      specMetadataPromise,
    ]);

    return {
      apiName,
      clientAgent,
      specificationId,
      sessionId: uuidv4(),
    };
  }
);

const initialState = {
  loading: true,
} as AsyncStatus<
  {
    apiName: string;
    clientAgent: string;
    specificationId: string;
    sessionId: string;
  },
  SerializedError
>;

const metadataSlice = createSlice({
  name: 'metadata',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(fetchMetadata.fulfilled, (state, action) => {
      return {
        loading: false,
        data: action.payload,
      };
    });
    builder.addCase(fetchMetadata.rejected, (state, action) => {
      return {
        loading: false,
        error: action.error,
      };
    });
  },
});

export const actions = {
  ...metadataSlice.actions,
  fetchMetadata,
};
export const reducer = metadataSlice.reducer;
