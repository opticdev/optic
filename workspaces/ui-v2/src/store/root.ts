import { configureStore } from '@reduxjs/toolkit';
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';

import { reducer as endpointsReducer } from './endpoints';
import { reducer as documentationEditReducer } from './documentationEdit';
import { reducer as metadataReducer } from './metadata';

export const createReduxStore = () =>
  configureStore({
    reducer: {
      endpoints: endpointsReducer,
      documentationEdits: documentationEditReducer,
      metadata: metadataReducer,
    },
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware({
        serializableCheck: false,
      }),
  });

export const store = createReduxStore();

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
