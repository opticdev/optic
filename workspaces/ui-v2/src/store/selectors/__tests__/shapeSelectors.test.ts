import { RootState } from '<src>/store/root';
import {
  getShapeRenderer,
  createFlatList,
  convertShapeToQueryParameters,
} from '../shapeSelectors';

import { getMockReduxStore } from './testHelpers';

let store: RootState;
beforeEach(() => {
  store = getMockReduxStore();
});
describe('getShapeRenderer', () => {
  test('loading', () => {
    store.shapes.rootShapes['shape_123'] = {
      loading: true,
    };
    expect(getShapeRenderer('shape_123')(store)).toEqual({
      loading: true,
    });
  });

  test('no shape', () => {
    expect(getShapeRenderer('shape_123')(store)).toEqual({
      loading: true,
    });
  });

  test('error', () => {
    store.shapes.rootShapes['shape_123'] = {
      loading: false,
      error: new Error('asd'),
    };
    const shapeState = getShapeRenderer('shape_123')(store);
    expect(shapeState.loading).toBe(false);
    expect(shapeState.error).toBeTruthy();
  });

  test('converts shapes to shape renderers', () => {
    const shapeState = getShapeRenderer('shape_cEkQAVQ3ib')(store);
    expect(shapeState.loading).toBe(false);
    expect(shapeState.data).toMatchSnapshot();
  });

  test('converts shapes to shape renderers with changelog', () => {
    const shapeState = getShapeRenderer(
      'shape_cEkQAVQ3ib',
      'batchcommitid'
    )(store);
    expect(shapeState.loading).toBe(false);
    expect(shapeState.data).toMatchSnapshot();
  });
});

describe('createFlatList', () => {
  test('converts shape to a contribution list', () => {
    const shapeState = getShapeRenderer('shape_cEkQAVQ3ib')(store);
    expect(
      createFlatList(shapeState.data!, 'path_UOIsxzICu5.GET')
    ).toMatchSnapshot();
  });
});

describe('convertShapeToQueryParameters', () => {
  test('converts a shape renderer into a query parameter', () => {
    const shapeState = getShapeRenderer('shape_tNRgroSwLj')(store);
    expect(convertShapeToQueryParameters(shapeState.data!)).toMatchSnapshot();
  });
});
