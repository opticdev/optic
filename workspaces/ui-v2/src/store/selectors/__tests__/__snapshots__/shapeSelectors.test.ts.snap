// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertShapeToQueryParameters converts a shape renderer into a query parameter 1`] = `
Object {
  "author": Object {
    "changes": null,
    "contributions": Object {},
    "fieldId": "field_hnI7P1UdbB",
    "isRemoved": false,
    "name": "author",
    "required": true,
    "shapeChoices": Array [
      Object {
        "jsonType": "String",
        "shapeId": "shape_qQT0krhOKn",
      },
    ],
    "shapeId": "shape_qQT0krhOKn",
  },
  "status": Object {
    "changes": "added",
    "contributions": Object {},
    "fieldId": "field_W8dnCvOHU",
    "name": "status",
    "required": true,
    "shapeChoices": Array [
      Object {
        "jsonType": "String",
        "shapeId": "shape_ZL8uOU2HQF",
      },
    ],
    "shapeId": "shape_ZL8uOU2HQF",
  },
}
`;

exports[`createFlatList converts shape to a contribution list 1`] = `
Array [
  Object {
    "contribution": Object {
      "contributionKey": "description",
      "endpointId": "path_UOIsxzICu5.GET",
      "id": "field_LRYtHDYkVO",
      "value": "",
    },
    "depth": 1,
    "name": "task",
    "shapes": Array [
      Object {
        "jsonType": "String",
        "shapeId": "shape_tUQxsgursF",
      },
    ],
  },
  Object {
    "contribution": Object {
      "contributionKey": "description",
      "endpointId": "path_UOIsxzICu5.GET",
      "id": "field_XM7KRqWOlV",
      "value": "",
    },
    "depth": 1,
    "name": "isDone",
    "shapes": Array [
      Object {
        "jsonType": "Boolean",
        "shapeId": "shape_9cUoBYpjJU",
      },
    ],
  },
  Object {
    "contribution": Object {
      "contributionKey": "description",
      "endpointId": "path_UOIsxzICu5.GET",
      "id": "field_9mczOWgNnu",
      "value": "",
    },
    "depth": 1,
    "name": "id",
    "shapes": Array [
      Object {
        "jsonType": "String",
        "shapeId": "shape_R4cTQ1zpOs",
      },
    ],
  },
]
`;

exports[`getShapeRenderer converts shapes to shape renderers 1`] = `
Array [
  Object {
    "asArray": Object {
      "shapeChoices": Array [
        Object {
          "asObject": Object {
            "fields": Array [
              Object {
                "changes": null,
                "contributions": Object {},
                "fieldId": "field_LRYtHDYkVO",
                "isRemoved": false,
                "name": "task",
                "required": true,
                "shapeChoices": Array [
                  Object {
                    "jsonType": "String",
                    "shapeId": "shape_tUQxsgursF",
                  },
                ],
                "shapeId": "shape_tUQxsgursF",
              },
              Object {
                "changes": null,
                "contributions": Object {},
                "fieldId": "field_XM7KRqWOlV",
                "isRemoved": false,
                "name": "isDone",
                "required": true,
                "shapeChoices": Array [
                  Object {
                    "jsonType": "Boolean",
                    "shapeId": "shape_9cUoBYpjJU",
                  },
                ],
                "shapeId": "shape_9cUoBYpjJU",
              },
              Object {
                "changes": "updated",
                "contributions": Object {},
                "fieldId": "field_9mczOWgNnu",
                "isRemoved": false,
                "name": "id",
                "required": true,
                "shapeChoices": Array [
                  Object {
                    "jsonType": "String",
                    "shapeId": "shape_R4cTQ1zpOs",
                  },
                ],
                "shapeId": "shape_R4cTQ1zpOs",
              },
            ],
          },
          "jsonType": "Object",
          "shapeId": "shape_f7gQgQ8p7G",
        },
      ],
      "shapeId": "shape_f7gQgQ8p7G",
    },
    "jsonType": "Array",
    "shapeId": "shape_cEkQAVQ3ib",
  },
]
`;

exports[`getShapeRenderer converts shapes to shape renderers with changelog 1`] = `
Array [
  Object {
    "asArray": Object {
      "shapeChoices": Array [
        Object {
          "asObject": Object {
            "fields": Array [
              Object {
                "changes": null,
                "contributions": Object {},
                "fieldId": "field_LRYtHDYkVO",
                "isRemoved": false,
                "name": "task",
                "required": true,
                "shapeChoices": Array [
                  Object {
                    "jsonType": "String",
                    "shapeId": "shape_tUQxsgursF",
                  },
                ],
                "shapeId": "shape_tUQxsgursF",
              },
              Object {
                "changes": null,
                "contributions": Object {},
                "fieldId": "field_XM7KRqWOlV",
                "isRemoved": false,
                "name": "isDone",
                "required": true,
                "shapeChoices": Array [
                  Object {
                    "jsonType": "Boolean",
                    "shapeId": "shape_9cUoBYpjJU",
                  },
                ],
                "shapeId": "shape_9cUoBYpjJU",
              },
              Object {
                "changes": "updated",
                "contributions": Object {},
                "fieldId": "field_9mczOWgNnu",
                "isRemoved": false,
                "name": "id",
                "required": true,
                "shapeChoices": Array [
                  Object {
                    "jsonType": "String",
                    "shapeId": "shape_R4cTQ1zpOs",
                  },
                ],
                "shapeId": "shape_R4cTQ1zpOs",
              },
              Object {
                "changes": "removed",
                "contributions": Object {},
                "fieldId": "field_5GCvc8KB2p",
                "isRemoved": true,
                "name": "dueDate",
                "required": true,
                "shapeChoices": Array [
                  Object {
                    "jsonType": "String",
                    "shapeId": "shape_owJFnZQJeS",
                  },
                ],
                "shapeId": "shape_owJFnZQJeS",
              },
            ],
          },
          "jsonType": "Object",
          "shapeId": "shape_f7gQgQ8p7G",
        },
      ],
      "shapeId": "shape_f7gQgQ8p7G",
    },
    "jsonType": "Array",
    "shapeId": "shape_cEkQAVQ3ib",
  },
]
`;
